[bmalloc] Prepare for InstallAPI by fixing headers and hiding libpas declarations

emw-apple opened the pull request:
#### c885def8396f6ffe1bd6cad5329cd22b62db2941
<pre>
[bmalloc] Prepare for InstallAPI by fixing headers and hiding libpas declarations
<a href="https://bugs.webkit.org/show_bug.cgi?id=254612">https://bugs.webkit.org/show_bug.cgi?id=254612</a>

Reviewed by Alexey Proskuryakov.

As part of the larger project of JavaScriptCore InstallAPI on Apple
platforms, make non-functional changes that clean up symbol visibility
inconsistencies in bmalloc headers, so that declared visibility matches
actual visibility.

To make this task simpler, start by refactoring libpas&apos;s symbols to be
hidden above JavaScriptCore:

1. The (C++) memory API used above JavaScriptCore is pure bmalloc, so in
   WebCore and WebKit, it is reasonable to treat libpas as a bmalloc
   implementation detail. Refactor bmalloc.xcodeproj to build libpas as
   a separate library target to separate its executable content from
   bmalloc.

2. JavaScriptCore needs to call into libpas, but the symbols can be
   hidden from its clients. Link against libpas.a using -hidden-lpas,
   which hides all its symbols.

3. Including bmalloc headers transitively includes libpas. To prevent
   JavaScriptCore&apos;s upcoming InstallAPI step from seeing
   externally-visible libpas declarations, JavaScriptCore compiles with
   a -DPAS_BMALLOC_HIDDEN flag which changes the PAS_API directive to
   hide declarations.

With libpas hidden, fix up a few visibility issues in bmalloc.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Link
  against libpas.a and compile with -DPAS_BMALLOC_HIDDEN, as described
  above.

* Source/bmalloc/Configurations/libpas.xcconfig: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
  - Create new libpas.a static library target. Move libpas sources over
    to this target (lots of PBXBuildFile churn).
  - Mark header visibility &quot;project-only&quot; for some headers that do not
    expose API. For the most part, these are headers which are only
    included by c/cpp files. They are no longer installed to
    /usr/local/include/bmalloc, and will not be processed by TAPI.

* Source/bmalloc/bmalloc/CryptoRandom.cpp: The ARC4RandomNumberGenerator
  storage is intentionally hidden. Use
  DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE to mark it BNOEXPORT.
* Source/bmalloc/bmalloc/Gigacage.cpp: The PrimitiveDisableCallbacks
  storage is intentionally hidden. Use
  DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE to mark it BNOEXPORT.
* Source/bmalloc/bmalloc/StaticPerProcess.h: Add
  DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE, to support declaring
  storage without forcing it to be BEXPORT.

* Source/bmalloc/libpas/src/libpas/pas_debug_heap.h: For the symbols
  that are implemented in bmalloc, use BEXPORT instead of PAS_API.
* Source/bmalloc/libpas/src/libpas/pas_enumerate_initially_unaccounted_pages.h:
  Add a missing PAS_END_EXTERN_C.
* Source/bmalloc/libpas/src/libpas/pas_enumerate_unaccounted_pages_as_meta.h:
  Add a missing PAS_END_EXTERN_C.
* Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h: Recognize a new
  PAS_BMALLOC_HIDDEN flag, which hides PAS_API symbols while still
  building like PAS_BMALLOC.

Canonical link: <a href="https://commits.webkit.org/262876@main">https://commits.webkit.org/262876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad9f0b8e70cf91385c91a5285ee85fc84c187f3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2876 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2965 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4219 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3246 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2773 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2468 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2834 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3241 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3999 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/736 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2488 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2356 "147 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2333 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2492 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2539 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3743 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2673 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2888 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2305 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2877 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2538 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2491 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/684 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/711 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2528 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2948 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2709 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/807 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_debug_heap.h:
```diff
@@ -38,12 +38,14 @@ PAS_BEGIN_EXTERN_C;
 
 #if PAS_BMALLOC
 
+#include "BExport.h"
```
bmalloc includes libpas. But libpas does not include bmalloc. Thus, we cannot include bmalloc header from libpas.

emw-apple replied:
libpas already does this:
https://github.com/WebKit/WebKit/blob/cffbf9134817e83c106e9303a9bd59ced78f7335/Source/bmalloc/libpas/src/libpas/pas_config.h#L33-L34
https://github.com/WebKit/WebKit/blob/cffbf9134817e83c106e9303a9bd59ced78f7335/Source/bmalloc/libpas/src/libpas/jit_heap_config.c#L43-L44

Since these includes are always guarded by `#if PAS_BMALLOC`, I think it's safe. In Xcode, these headers are always available (because libpas and bmalloc are the same project).

If we should avoid adding additional imports like this, I could inline the macro. Or maybe these declarations can be moved around. What do you think?

aproskuryakov replied:
Please add a FIXME about this.

aproskuryakov commented on Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
```diff
@@ -64,7 +64,7 @@ CLANG_INSTRUMENT_FOR_OPTIMIZATION_PROFILING = $(CLANG_INSTRUMENT_FOR_OPTIMIZATIO
 
 GCC_PREFIX_HEADER = JavaScriptCorePrefix.h;
-GCC_PREPROCESSOR_DEFINITIONS = $(inherited) $(ENABLE_LLVM_PROFILE_GENERATION);
+GCC_PREPROCESSOR_DEFINITIONS = $(inherited) $(ENABLE_LLVM_PROFILE_GENERATION) PAS_BMALLOC_HIDDEN;
```
In WebKit, we have the `BUILDING_WEBKIT` macro, which is one way to make headers behave differently when building the library than when using it. I don't have a string opinion, but wanted to suggest `BUILDING_JAVASCRIPTCORE` as a potential alternative to `PAS_BMALLOC_HIDDEN`.

...actually we also have `BUILDING_WebKit` with a different capitalization...

emw-apple replied:
I wanted to avoid a layering violation—why should bmalloc headers behave in a unique way when JavaScriptCore is building?—but if you think that's better than a bespoke new configuration macro, I'm happy to do it.

aproskuryakov replied:
I don't have a preference, just wanted to point out an alternative to consider.

aproskuryakov commented on Source/bmalloc/bmalloc/CryptoRandom.cpp:
```diff
@@ -75,7 +75,7 @@ class ARC4RandomNumberGenerator : public StaticPerProcess<ARC4RandomNumberGenera
     int m_count;
 };
-DECLARE_STATIC_PER_PROCESS_STORAGE(ARC4RandomNumberGenerator);
+DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE(ARC4RandomNumberGenerator, BNOEXPORT);
```
This looks scary - how do we enforce that it's per process if we don't export it?

If this is how we have to do it, that needs at least a comment explaining why it doesn't cause problems.

emw-apple replied:
I think the key is that this storage is only known to this source file. So it's not possible (without eyebrow-raising forward declarations) to access it outside this compilation unit.

aproskuryakov replied:
Is this entirely correct? The "per process" naming implies to me that we don't want two of those to be brought in into the process. Which seems easier to do accidentally now.

emw-apple replied:
Maybe a better way to put this is: `ARC4RandomNumberGenerator` an implementation detail of `bmalloc::cryptoRandom()`. It's declared in this cpp file and is not exported. So it would not be possible to have a declaration of this class and its inlined storage-accessing methods outside of the compilation unit. This change gives the storage object the same visibility as the class itself.

aproskuryakov commented on Source/bmalloc/bmalloc/Gigacage.cpp:
```diff
@@ -73,7 +73,7 @@ struct PrimitiveDisableCallbacks : public StaticPerProcess<PrimitiveDisableCallb
     Vector<Gigacage::Callback> callbacks;
 };
-DECLARE_STATIC_PER_PROCESS_STORAGE(PrimitiveDisableCallbacks);
+DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE(PrimitiveDisableCallbacks, BNOEXPORT);
```
Ditto.

emw-apple replied:
Same as above: `PrimitiveDisableCallbacks` is an implementation detail of the primitive Gigacage APIs declared in "Gigacage.h". This change gives the storage object the same visibility as the class itself.

aproskuryakov commented on Source/bmalloc/libpas/src/libpas/pas_debug_heap.h:
```diff
@@ -38,12 +38,14 @@ PAS_BEGIN_EXTERN_C;
+
 /* The implementations are provided by bmalloc. */
-PAS_API extern bool pas_debug_heap_is_enabled(pas_heap_config_kind kind);
-PAS_API extern void* pas_debug_heap_malloc(size_t size);
-PAS_API extern void* pas_debug_heap_memalign(size_t alignment, size_t size);
-PAS_API extern void* pas_debug_heap_realloc(void* ptr, size_t size);
-PAS_API extern void pas_debug_heap_free(void* ptr);
+BEXPORT extern bool pas_debug_heap_is_enabled(pas_heap_config_kind);
+BEXPORT extern void* pas_debug_heap_malloc(size_t);
+BEXPORT extern void* pas_debug_heap_memalign(size_t alignment, size_t);
+BEXPORT extern void* pas_debug_heap_realloc(void* ptr, size_t);
+BEXPORT extern void pas_debug_heap_free(void* ptr);
```
The separation between `PAS` and `B` seems generally confusing. Not sure if there's anything actionable or blocking there.

emw-apple replied:
As I understand it, the three different linkage macros work like this:

- `PAS_API`: core libpas symbols, visible except when libpas is configured to build as an opaque `libmalloc` (afaik, WebKit does not ever build libpas in this mode)
- `PAS_BAPI`: symbols that are only visible when libpas is configured to integrate with bmalloc
- `BEXPORT`: bmalloc's export macro, visible when building bmalloc and otherwise hidden

These symbols are a fun edge case where libpas forward declares bmalloc symbols it, to avoid a cycle. The comment above reads,

```
/* Bmalloc has a DebugHeap singleton that can be used to divert bmalloc calls to system malloc.
   When libpas is used in bmalloc, we use this to glue libpas into that mechanism. */
```

aproskuryakov commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -1053,6 +1053,8 @@
 		D04CF93F285C77CA005D6337 /* MachSendRight.cpp in Sources */ = {isa = PBXBuildFile; fileRef = D04CF93E285C77C9005D6337 /* MachSendRight.cpp */; };
 		DD0EDF8D2798A6AD005152AD /* libgtest.a in Product Dependencies */ = {isa = PBXBuildFile; fileRef = F3FC3EE213678B7300126A65 /* libgtest.a */; };
+		DD28DF2729D41E8100BBB459 /* libpas.a in Frameworks */ = {isa = PBXBuildFile; fileRef = DD28DF2629D41E8100BBB459 /* libpas.a */; };
+		DD28DF2F29D41E8F00BBB459 /* libpas.a in Frameworks */ = {isa = PBXBuildFile; fileRef = DD28DF2629D41E8100BBB459 /* libpas.a */; };
```
Duplicated lines.

emw-apple committed:
[bmalloc] Prepare for InstallAPI by fixing headers and hiding libpas declarations
https://bugs.webkit.org/show_bug.cgi?id=254612

Reviewed by Alexey Proskuryakov.

As part of the larger project of JavaScriptCore InstallAPI on Apple
platforms, make non-functional changes that clean up symbol visibility
inconsistencies in bmalloc headers, so that declared visibility matches
actual visibility.

To make this task simpler, start by refactoring libpas's symbols to be
hidden above JavaScriptCore:

1. The (C++) memory API used above JavaScriptCore is pure bmalloc, so in
   WebCore and WebKit, it is reasonable to treat libpas as a bmalloc
   implementation detail. Refactor bmalloc.xcodeproj to build libpas as
   a separate library target to separate its executable content from
   bmalloc.

2. JavaScriptCore needs to call into libpas, but the symbols can be
   hidden from its clients. Link against libpas.a using -hidden-lpas,
   which hides all its symbols.

3. Including bmalloc headers transitively includes libpas. To prevent
   JavaScriptCore's upcoming InstallAPI step from seeing
   externally-visible libpas declarations, JavaScriptCore compiles with
   a -DPAS_BMALLOC_HIDDEN flag which changes the PAS_API directive to
   hide declarations.

With libpas hidden, fix up a few visibility issues in bmalloc.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig: Link
  against libpas.a and compile with -DPAS_BMALLOC_HIDDEN, as described
  above.

* Source/bmalloc/Configurations/libpas.xcconfig: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
  - Create new libpas.a static library target. Move libpas sources over
    to this target (lots of PBXBuildFile churn).
  - Mark header visibility "project-only" for some headers that do not
    expose API. For the most part, these are headers which are only
    included by c/cpp files. They are no longer installed to
    /usr/local/include/bmalloc, and will not be processed by TAPI.

* Source/bmalloc/bmalloc/CryptoRandom.cpp: The ARC4RandomNumberGenerator
  storage is intentionally hidden. Use
  DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE to mark it BNOEXPORT.
* Source/bmalloc/bmalloc/Gigacage.cpp: The PrimitiveDisableCallbacks
  storage is intentionally hidden. Use
  DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE to mark it BNOEXPORT.
* Source/bmalloc/bmalloc/StaticPerProcess.h: Add
  DECLARE_STATIC_PER_PROCESS_STORAGE_WITH_LINKAGE, to support declaring
  storage without forcing it to be BEXPORT.

* Source/bmalloc/libpas/src/libpas/pas_debug_heap.h: For the symbols
  that are implemented in bmalloc, use BEXPORT instead of PAS_API.
* Source/bmalloc/libpas/src/libpas/pas_enumerate_initially_unaccounted_pages.h:
  Add a missing PAS_END_EXTERN_C.
* Source/bmalloc/libpas/src/libpas/pas_enumerate_unaccounted_pages_as_meta.h:
  Add a missing PAS_END_EXTERN_C.
* Source/bmalloc/libpas/src/libpas/pas_utils_prefix.h: Recognize a new
  PAS_BMALLOC_HIDDEN flag, which hides PAS_API symbols while still
  building like PAS_BMALLOC.

Canonical link: https://commits.webkit.org/262876@main
